### PR TITLE
Fixed #29689 -- Improved performance of FileSystemStorage.listdir() and FilePathField with os.scandir().

### PR DESCRIPTION
**Pull request by federicobond**
https://code.djangoproject.com/ticket/29689

### CODE REVIEW

**Comment by timgraham** [comment]
Did you notice the test failures?

**Comment by federicobond** [comment]
I hadn’t, thanks for amending it. Looks like it passes now.

**Comment by timgraham** [comment]
Yep, just needed to change `entry.path` to `entry.name`. Do you have any benchmarks to prove that this improves performance? Did you audit the other `os.path.list()` calls in Django and confirm that they should remain?

**Comment by federicobond** [comment]
* I noticed a substantial speedup while iterating over thousands of files (+50%). The performance advantages of scandir are well documented in [PEP 471](https://www.python.org/dev/peps/pep-0471/), as it avoids an additional `stat` syscall for every entry. If you like, I can isolate that code and do some proper benchmarks.
 * There is another use of `os.listdir` in `django.forms.fields.FilePathField` which could be improved by replacing it with `scandir`. I have already produced a patch but I would prefer to discuss it separately. Would you like me to bundle those changes here instead?

**Comment by timgraham** [comment]
I think it would be fine to include that change here, perhaps as a separate commit. To confirm my understanding, the additional syscall comes from calling `os.path.isdir()` for each entry, is that correct? If so, I guess it makes enough sense that we don't need formal benchmarks.

**Comment by federicobond** [comment]
Yes, it avoids an additional syscall from calling `os.path.isdir()` and also improves memory efficiency by returning an iterator instead of a list of entries.

**Comment by federicobond** [comment]
@timgraham I have added the code for FilePathField in another commit.

**Comment by federicobond** [comment]
@timgraham thanks for cleaning up my mess. I was having trouble running those tests locally.

File: django/forms/fields.py [hunk 1]
```diff
@@ -1104,17 +1104,20 @@ def __init__(self, path, *, match=None, recursive=False, allow_files=True,
...
-                            (self.allow_folders and os.path.isdir(full_file))) and
-                            (self.match is None or self.match_re.search(f))):
-                        self.choices.append((full_file, f))
+                    if (((self.allow_files and f.is_file()) or
+                            (self.allow_folders and f.is_dir())) and
+                            (self.match is None or self.match_re.search(f.name))):
+                        choices.append((f.path, f.name))
+                    choices.sort(key=lambda c: c[1])
             except OSError:
                 pass
```
Comment (federicobond) [inline django/forms/fields.py]: Now that there isn't a potential race condition between calling `listdir` and `isfile`/`isdir`, perhaps this try/except can be removed too.

File: django/forms/fields.py [hunk 2]
```diff
@@ -1104,17 +1104,20 @@ def __init__(self, path, *, match=None, recursive=False, allow_files=True,
...
-                    full_file = os.path.join(self.path, f)
-                    if (((self.allow_files and os.path.isfile(full_file)) or
-                            (self.allow_folders and os.path.isdir(full_file))) and
-                            (self.match is None or self.match_re.search(f))):
-                        self.choices.append((full_file, f))
+                    if (((self.allow_files and f.is_file()) or
+                            (self.allow_folders and f.is_dir())) and
+                            (self.match is None or self.match_re.search(f.name))):
+                        choices.append((f.path, f.name))
+                    choices.sort(key=lambda c: c[1])
```
Comment (charettes) [inline django/forms/fields.py]: I guess `operator.itemgetter(1)` could be used here.

**Pull request merged**